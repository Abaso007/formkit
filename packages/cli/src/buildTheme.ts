import { existsSync } from 'fs'
import { writeFile } from 'fs/promises'
import { resolve } from 'pathe'
import { error, green, __dirname, info } from './index'
import { Theme, ThemeOptions } from '@formkit/theme-creator'
import { stylesheetFromTailwind } from '@formkit/theme-creator/stylesheet'

interface BuildThemeOptions {
  semantic: boolean
  variables?: string
  api?: string
  format?: 'ts' | 'mjs'
  outFile?: string
}

const DEFAULT_THEME_API = 'https://themes.formkit.com/api'
const HAS_EXTENSION_RE = /\.(?:ts|js|mjs|cjs)$/

export async function buildTheme(
  themeName: string,
  options: Partial<BuildThemeOptions> = {}
) {
  if (!themeName) error('Please provide a theme name or path to a theme file.')
  green(`Locating ${themeName}...`)
  const endpoint = options.api || DEFAULT_THEME_API

  const format = options.format || guessFormat()
  const theme = themeName?.startsWith('./')
    ? await generate(
        await localTheme(themeName),
        options.variables,
        format === 'ts',
        options.semantic
      )
    : await apiTheme(themeName, endpoint)

  const outFile =
    options.outFile || 'formkit.theme.' + (options.semantic ? 'css' : format)
  await writeFile(resolve(process.cwd(), outFile), theme)
  green(`Theme file written to ${outFile}`)
}

function guessFormat() {
  return existsSync(resolve(process.cwd(), 'tsconfig.json')) ? 'ts' : 'mjs'
}

async function generate(
  theme: Theme<ThemeOptions>,
  variables?: string,
  isTS?: boolean,
  semantic?: boolean
): Promise<string> {
  info(`Loaded theme: ${theme.meta.name}`)
  const vars = parseVariables(variables)
  const classes = theme(vars).tailwind()
  if (semantic) return await stylesheetFromTailwind(classes)
  const classList: Record<string, Record<string, true>> = {}
  const globals: Record<string, Record<string, true>> = {}

  for (const input in classes) {
    for (const section in classes[input]) {
      const key = `${input}__${section}`
      const sectionClasses = classes[input][section]
        .split(' ')
        .reduce((acc, cur) => {
          acc[cur] = true
          return acc
        }, {} as Record<string, true>)
      if (input === '__globals') {
        globals[section] = sectionClasses
      } else {
        classList[key] = sectionClasses
      }
    }
  }

  const themeFile = `${
    isTS ? "import type { FormKitNode } from '@formkit/core'\n\n" : ''
  }/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(classList, null, 2)};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(globals, null, 2)};

/**
 * This is the theme function itself, it should be imported and used as the
 * config.rootClasses function. For example:
 *
 * \`\`\`js
 * import { theme } from './formkit.theme'
 * import { defineFormKitConfig } from '@formkit/vue'
 *
 * export default defineFormKitConfig({
 *   config: {
 *     rootClasses: theme
 *   }
 * })
 * \`\`\`
 **/
export function theme (sectionName${isTS ? ': string' : ''}, node${
    isTS ? ': FormKitNode' : ''
  })${isTS ? ': Record<string, boolean>' : ''} {
  const key = \`\${node.props.type}__\${sectionName}\`
  if (key in classes) return classes[key]
  if (sectionName in globals) return classes[sectionName]
  return {}
}
`
  return themeFile
}

function parseVariables(variables?: string): Record<string, string> {
  if (!variables) return {}
  return variables.split(',').reduce((vars, unparsed) => {
    const [key, value] = unparsed.split('=')
    vars[key] = value
    return vars
  }, {} as Record<string, string>)
}

function getPath(paths: string[]): string | undefined {
  const path = paths.shift()
  if (existsSync(path!)) return path
  return paths.length ? getPath(paths) : undefined
}

async function localTheme(
  themeName: string
): Promise<Theme<ThemeOptions>> | never {
  const extensions = ['.ts', '.js', '.mjs', '.cjs']
  const paths = HAS_EXTENSION_RE.test(themeName)
    ? [resolve(process.cwd(), themeName)]
    : extensions.map((ext) => resolve(process.cwd(), themeName + ext))
  const path = getPath(paths)
  if (!path) error(`Could not find ${themeName}.`)
  const theme = (await import(path)) as { default: Theme<ThemeOptions> }
  if (typeof theme !== 'object' || !theme.default) error('Invalid theme file.')
  return theme.default
}

async function apiTheme(themeName: string, endpoint: string): Promise<string> {
  return `${endpoint}${themeName}`
}
