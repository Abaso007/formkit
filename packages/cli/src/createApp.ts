import { execa, execaCommand } from 'execa'
import { readFile, writeFile, readdir } from 'fs/promises'
import { resolve } from 'path'
import { cwd } from 'node:process'
import prompts from 'prompts'
import { error, green, __dirname, info } from './index'
import ora from 'ora'
import http from 'http'
import url from 'url'
import open from 'open'

const APP_URL = 'https://pro.formkit.com'
interface CreateAppOptions {
  lang: 'ts' | 'js'
  framework: 'nuxt' | 'vite'
  pro?: string
}

async function login(): Promise<string> {
  const spinner = ora(`To login visit: ${APP_URL}/cli-login`).start()
  await open(`${APP_URL}/cli-login`)
  let server: http.Server | undefined
  const token = await new Promise<string>((resolve, reject) => {
    server = http
      .createServer((req, res) => {
        const urlObj = url.parse(req.url!, true)
        const token = urlObj.query.token as string | undefined

        if (token) {
          resolve(token)
          res.writeHead(302, {
            Location: `${APP_URL}/cli-login?success=true`,
          })
          res.end()
        } else {
          res.writeHead(302, {
            Location: `${APP_URL}/cli-login?success=false`,
          })
          reject('Login failed.')
        }
      })
      .listen(5479)
  })
  server?.close()
  spinner.stop()
  return token
}

async function createTeam(token: string) {
  const res = await prompts({
    type: 'text',
    name: 'name',
    message: 'Please enter a new team name:',
    initial: 'My team',
  })
  const response = await fetch(`${APP_URL}/api/teams`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify({
      name: res.name,
    }),
  })
  return await response.json()
}

async function createProject(token: string, team: number) {
  const res = await prompts({
    type: 'text',
    name: 'name',
    message: 'Please enter a new project name:',
    initial: 'My new project',
  })
  const response = await fetch(`${APP_URL}/api/teams/${team}/projects`, {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    },
    body: JSON.stringify({
      name: res.name,
      license: 'development',
    }),
  })
  if (!response.ok) {
    error('Failed to create project.')
  }
  const data = await response.json()
  if (data.data) {
    info(
      `Your project was created successfully with a development license ‚Äî to upgrade to a production license visit ${APP_URL}`
    )
  }
  return data.data
}

async function selectProProject() {
  try {
    const token = (await login()) as string
    const spinner = ora('Fetching account...').start()
    const response = await fetch(`${APP_URL}/api/account`, {
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    })
    const data = await response.json()
    spinner.stop()

    const res = await prompts({
      type: 'select',
      name: 'team',
      message: 'Select a team:',
      choices: data.teams
        .map(
          (team: {
            id: number
            name: string
            projects: Array<{ name: string; api_key: string }>
          }) => ({
            title: team.name,
            value: team,
          })
        )
        .concat([{ title: 'Create a new team', value: 'new' }]),
    })

    if (res.team === 'new') {
      const team = await createTeam(token)
      const project = await createProject(token, team.id)
      return project.api_key
    } else {
      let { project } = await prompts({
        type: 'select',
        name: 'project',
        message: 'Select a project:',
        choices: res.team.projects
          .map((team: { id: number; name: string }) => ({
            title: team.name,
            value: team,
          }))
          .concat([{ title: 'Create a new project', value: 'new' }]),
      })

      if (project === 'new') {
        project = await createProject(token, res.team.id)
      }
      return project.api_key
    }
  } catch (err) {
    console.log(err)
    error('Login failed.')
  }
  return 'test'
}

export async function createApp(
  appName?: string,
  options: Partial<CreateAppOptions> = {}
): Promise<void> {
  if (!appName) {
    const res = await prompts({
      type: 'text',
      name: 'name',
      message: 'Please enter a directory name for the project:',
      initial: 'formkit-app',
    })
    appName = res.name as string
  }

  const isEmpty = await isDirEmpty(resolve(cwd(), `./${appName}`))
  if (!isEmpty) {
    error('Directory is not empty. Please choose a different name.')
  }

  if (!options.framework) {
    const res = await prompts({
      type: 'select',
      name: 'framework',
      message: 'What framework would you like to use?',
      choices: [
        { title: 'Vite', value: 'vite' },
        { title: 'Nuxt', value: 'nuxt' },
      ],
      initial: 1,
    })
    options.framework = res.framework as 'vite' | 'nuxt'
  }

  if (!options.lang && options.framework === 'vite') {
    const res = await prompts({
      type: 'select',
      name: 'lang',
      message: 'What language should be used?',
      choices: [
        { title: 'TypeScript', value: 'ts' },
        { title: 'JavaScript', value: 'js' },
      ],
      initial: 1,
    })
    options.lang = res.lang as 'ts' | 'js'
  }

  if (!options.pro) {
    const res = await prompts([
      {
        type: 'toggle',
        name: 'install_pro',
        message: 'Would you like to install FormKit Pro?',
        active: 'yes',
        initial: true,
        inactive: 'no',
      },
    ])
    if (res.install_pro) {
      options.pro = await selectProProject()
    }
  }

  if (options.framework === 'vite') {
    await execa('npx', [
      '--yes',
      'create-vite@latest',
      appName,
      '--template',
      `vue${options.lang === 'ts' ? '-ts' : ''}`,
    ])
    // TODO: add better version matching here:
    await addDependency(appName, '@formkit/vue', 'latest')
    await addDependency(appName, '@formkit/icons', 'latest')
    if (options.pro) {
      await addDependency(appName, '@formkit/pro')
    }
    await addInitialApp(appName, 'src/App.vue', !!options.pro)
    await writeFile(
      resolve(cwd(), `./${appName}/src/formkit.config.${options.lang}`),
      buildFormKitConfig(options as CreateAppOptions)
    )
    await writeFile(
      resolve(cwd(), `./${appName}/src/main.${options.lang}`),
      buildMain()
    )
  } else {
    options.lang = 'ts'
    info('Fetching nuxi cli...')
    const subprocess = execaCommand(
      `npx --yes nuxi@latest init --no-install $APP_NAME`,
      {
        cwd: process.cwd(),
        shell: true,
        stdio: 'inherit',
        env: { APP_NAME: appName },
      }
    )
    subprocess.stdout?.pipe(process.stdout)
    subprocess.stderr?.pipe(process.stderr)
    await subprocess
    await writeFile(
      resolve(cwd(), `./${appName}/formkit.config.ts`),
      buildFormKitConfig(options as CreateAppOptions)
    )
    await addDependency(appName, '@formkit/nuxt', 'latest')
    await addDependency(appName, '@formkit/icons', 'latest')
    if (options.pro) {
      await addDependency(appName, '@formkit/pro')
    }
    await addNuxtModule(appName)
    await addInitialApp(appName, 'app.vue', !!options.pro)
  }

  green(`Created ${appName}!

To run your new app:
üìÅ cd ${appName}
‚úÖ npm install
üöÄ npm run dev
`)
}

async function addInitialApp(dirName: string, component: string, pro: boolean) {
  const appPath = resolve(cwd(), `./${dirName}/${component}`)
  await writeFile(
    appPath,
    `<script setup>
async function submit() {
  await new Promise(r => setTimeout(r, 1000))
  alert('Submitted! üéâ')
}
</script>

<template>
  <div class="your-first-form">
    <img
      src="https://pro.formkit.com/logo.svg"
      alt="FormKit Logo"
      width="244"
      height="50"
      class="logo"
    >
    <FormKit
      type="form"
      #default="{ value }"
      @submit="submit"
    >
      <FormKit
        type="text"
        name="name"
        label="Name"
        help="What do people call you?"
      />
      <FormKit
        type="checkbox"
        name="flavors"
        label="Favorite ice cream flavors"
        :options="{
          'vanilla': 'Vanilla',
          'chocolate': 'Chocolate',
          'strawberry': 'Strawberry',
          'mint-chocolate-chip': 'Mint Chocolate Chip',
          'rocky-road': 'Rocky Road',
          'cookie-dough': 'Cookie Dough',
          'pistachio': 'Pistachio',
        }"
        validation="required|min:2"
      />
      ${
        (pro &&
          `
      <FormKit
        type="repeater"
        name="invitees"
        label="Invitees"
        help="Who else should we invite to FormKit?"
      >
        <FormKit
          type="text"
          name="email"
          label="Email"
          validation="required|email"
        />
      </FormKit>`) ||
        ''
      }
      <FormKit
        type="checkbox"
        name="agree"
        label="I agree FormKit is the best form authoring framework."
      />
      <pre>{{ value }}</pre>
    </FormKit>
  </div>
</template>

<style scoped>
.your-first-form {
  width: calc(100% - 2em);
  max-width: 480px;
  box-sizing: border-box;
  padding: 2em;
  box-shadow: 0 0 1em rgba(0, 0, 0, .1);
  border-radius: .5em;
  margin: 4em auto;
}

.logo {
  width: 150px;
  height: auto;
  display: block;
  margin: 0 auto 2em auto;
}
pre {
  background-color: rgba(0, 100, 250, .1);
  padding: 1em;
}
</style>
`
  )
}

/**
 * Adds a dependency to a new project‚Äôs package.json file.
 * @param dirName - The directory to find a package.json
 * @param dependency - An npm dependency to add.
 */
async function addDependency(
  dirName: string,
  dependency: string,
  version = 'latest'
) {
  const packageJsonPath = resolve(cwd(), `./${dirName}/package.json`)
  const raw = await readFile(packageJsonPath, 'utf-8')
  const packageJson = JSON.parse(raw)
  if (!('dependencies' in packageJson)) {
    packageJson.dependencies = {}
  }
  packageJson.dependencies[dependency] = version
  await writeFile(packageJsonPath, JSON.stringify(packageJson, null, 2))
}

function buildMain() {
  return `import { createApp } from 'vue'
import { plugin, defaultConfig } from '@formkit/vue'
import App from './App.vue'
import formKitConfig from './formkit.config'

const app = createApp(App)
app.use(plugin, defaultConfig(formKitConfig))
app.mount('#app')
`
}

async function addNuxtModule(dirName: string) {
  const nuxtConfigPath = resolve(cwd(), `./${dirName}/nuxt.config.ts`)
  const raw = await readFile(nuxtConfigPath, 'utf-8')
  const configWithFormKit = raw.replace(
    /(defineNuxtConfig\({\n).*?(\n}\))/g,
    "$1  modules: ['@formkit/nuxt'],\n  formkit: {\n    autoImport: true\n  }$2"
  )
  await writeFile(nuxtConfigPath, configWithFormKit)
}

/**
 * Builds the formkit.config.ts file.
 * @param options - Build the formkit.config.ts file for a Nuxt project.
 * @returns
 */
function buildFormKitConfig(options: CreateAppOptions): string {
  const imports = [
    'import "@formkit/themes/genesis"',
    'import { genesisIcons } from "@formkit/icons"',
  ]
  if (options.lang === 'ts' && options.framework === 'vite') {
    imports.push("import { DefaultConfigOptions } from '@formkit/vue'")
  } else if (options.lang === 'ts' && options.framework === 'nuxt') {
    imports.push("import { defineFormKitConfig } from '@formkit/vue'")
  }
  const setup = []
  let config = ''
  if (options.pro) {
    imports.push("import { createProPlugin, inputs } from '@formkit/pro'")
    imports.push("import '@formkit/pro/genesis'")
    setup.push('')
    setup.push(`const pro = createProPlugin('${options.pro}', inputs)`)
    setup.push('')
    config += `  plugins: [pro]`
  }
  config += `${config ? ',\n' : ''}  icons: { ...genesisIcons }`

  const viteExport = `const config${
    options.lang === 'ts' ? ': DefaultConfigOptions' : ''
  } = {
${config}
}

export default config`

  const nuxtExport = `export default defineFormKitConfig({
${config}
})`

  let defaultExport = ''
  if (options.framework === 'nuxt') {
    defaultExport = nuxtExport
  } else if (options.framework === 'vite') {
    defaultExport = viteExport
  }

  const rawConfig = `${imports.join('\n')}
${setup.join('\n')}
${defaultExport}
`

  return rawConfig
}

async function isDirEmpty(path: string) {
  try {
    const entries = await readdir(path)
    return entries.length === 0
  } catch (error) {
    return true
  }
}
