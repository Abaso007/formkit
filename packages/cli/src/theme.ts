import { existsSync } from 'fs'
import { writeFile } from 'fs/promises'
import { resolve } from 'pathe'
import { error, green, __dirname, info } from './index'
import { Theme, ThemeOptions } from '@formkit/theme-creator'
import { stylesheetFromTailwind } from '@formkit/theme-creator/stylesheet'
import prompts from 'prompts'
import { createHash } from 'crypto'

interface BuildThemeOptions {
  semantic: boolean
  theme?: string
  variables?: string
  api?: string
  format?: 'ts' | 'mjs'
  outFile?: string
}

const DEFAULT_THEME_API = 'https://themes.formkit.com/api'
const HAS_EXTENSION_RE = /\.(?:ts|js|mjs|cjs)$/

export async function buildTheme(options: Partial<BuildThemeOptions> = {}) {
  let themeName = options.theme
  if (!options.theme) {
    const res = await fetch(`${DEFAULT_THEME_API}/themes`)
    const themes = await res.json()
    const { theme } = await prompts({
      type: 'select',
      message: 'Select a theme',
      name: 'theme',
      choices: themes.map((theme: any) => ({
        title: theme.name,
        value: theme.slug,
        description:
          theme.description +
          (theme.darkMode ? ' üåú' : '') +
          (theme.lightMode ? '‚òÄÔ∏è' : '') +
          ')',
      })),
    })
    themeName = theme
  }
  if (!themeName) error('Please provide a theme name or path to a theme file.')
  green(`Locating ${themeName}...`)
  const endpoint = options.api || DEFAULT_THEME_API

  const format = options.format || guessFormat()
  const theme = themeName?.startsWith('./')
    ? await generate(
        await localTheme(themeName),
        options.variables,
        format === 'ts',
        options.semantic
      )
    : await apiTheme(
        themeName,
        endpoint,
        options.variables ?? '',
        format === 'ts',
        options.semantic ?? false
      )

  const outFile =
    options.outFile || 'formkit.theme.' + (options.semantic ? 'css' : format)
  await writeFile(resolve(process.cwd(), outFile), theme)
  green(`Theme file written to ${outFile}`)
}

function guessFormat() {
  return existsSync(resolve(process.cwd(), 'tsconfig.json')) ? 'ts' : 'mjs'
}

export async function generate(
  theme: Theme<ThemeOptions>,
  variables?: string,
  isTS?: boolean,
  semantic?: boolean
): Promise<string> {
  info(`Loaded theme: ${theme.meta.name}`)
  const vars = parseVariables(variables)
  const classes = theme(vars).tailwind()
  if (semantic) return await stylesheetFromTailwind(classes)
  const classList: Record<string, Record<string, true>> = {}
  const globals: Record<string, Record<string, true>> = {}

  for (const input in classes) {
    for (const section in classes[input]) {
      const key = `${input}__${section}`
      const sectionClasses = classes[input][section]
        .split(' ')
        .reduce((acc, cur) => {
          acc[cur] = true
          return acc
        }, {} as Record<string, true>)
      if (input === '__globals') {
        globals[section] = sectionClasses
      } else {
        classList[key] = sectionClasses
      }
    }
  }

  const themeFile = `${
    isTS ? "import type { FormKitNode } from '@formkit/core'\n\n" : ''
  }/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you‚Äôd like to "eject" from the CLI‚Äôs ability to update it.
  *
  * @checksum -
  * @variables - ${variables}
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName${isTS ? ': string' : ''}, node${
    isTS ? ': FormKitNode' : ''
  })${isTS ? ': Record<string, boolean>' : ''} {
   const key = \`\${node.props.type}__\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey]
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(classList, null, 2)};

/**
 * Globals are merged prior to generating this file ‚Äî¬†these are included for
 * any other non-matching inputs.
 **/
const globals${
    isTS ? ': Record<string, Record<string, boolean>>' : ''
  } = ${JSON.stringify(globals, null, 2)};
`

  const checksum = createHash('sha256').update(themeFile).digest('hex')
  console.log({ checksum })
  return themeFile.replace(/@checksum -/, `@checksum - ${checksum}`)
}

function parseVariables(variables?: string): Record<string, string> {
  if (!variables) return {}
  return variables.split(',').reduce((vars, unparsed) => {
    const [key, value] = unparsed.split('=')
    vars[key] = value
    return vars
  }, {} as Record<string, string>)
}

function getPath(paths: string[]): string | undefined {
  const path = paths.shift()
  if (existsSync(path!)) return path
  return paths.length ? getPath(paths) : undefined
}

async function localTheme(
  themeName: string
): Promise<Theme<ThemeOptions>> | never {
  const extensions = ['.ts', '.js', '.mjs', '.cjs']
  const paths = HAS_EXTENSION_RE.test(themeName)
    ? [resolve(process.cwd(), themeName)]
    : extensions.map((ext) => resolve(process.cwd(), themeName + ext))
  const path = getPath(paths)
  if (!path) error(`Could not find ${themeName}.`)

  const theme = (await import(path)) as { default: Theme<ThemeOptions> }
  if (typeof theme !== 'object' || !theme.default) error('Invalid theme file.')
  return theme.default
}

async function apiTheme(
  themeName: string,
  endpoint: string,
  variables: string,
  isTS: boolean,
  semantic: boolean
): Promise<string> {
  info(`Generating theme: ${themeName}`)
  const res = await fetch(`${endpoint}/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      theme: themeName,
      variables,
      ts: String(isTS),
      semantic: String(semantic),
    }),
  })
  if (res.ok) {
    const code = await res.text()
    return code
  } else {
    error(`Could not generate theme ‚Äî ${res.statusText}`)
  }
}
