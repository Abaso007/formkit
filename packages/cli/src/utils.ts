import { readdir } from 'fs/promises'

/**
 * Checks if a given directory is empty.
 * @param path - The absolute path to check if empty.
 */
export async function isDirEmpty(path: string) {
  try {
    const entries = await readdir(path)
    return entries.length === 0
  } catch (error) {
    return true
  }
}
