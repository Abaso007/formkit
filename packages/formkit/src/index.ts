console.warn(
  'The npm package "formkit" is a command line utility. Please use @formkit/vue instead.'
)
export {}
