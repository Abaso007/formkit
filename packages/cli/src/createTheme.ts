import prompts from 'prompts'
import { downloadTemplate } from 'giget'
import { slugify } from '@formkit/utils'
import { isDirEmpty } from './utils'
import { resolve } from 'pathe'
import ora from 'ora'

interface CreateThemeOptions {
  name: string
  packageName: string
  dir: string
}

export async function createTheme(
  name: string,
  initialOptions: Partial<CreateThemeOptions>
) {
  // Do some stuff here
  const settings = await completeOptions(name, initialOptions)
  const spinner = ora(`Downloading starter theme...`).start()
  await downloadTemplate('github:formkit/theme-starter', {
    dir: settings.dir,
  })
  spinner.text = `Customizing package.json...`
  await customizePackageJson(settings)
  spinner.stop()
}

async function customizePackageJson(_options: CreateThemeOptions) {
  return new Promise((resolve) => setTimeout(resolve, 1000))
}

async function completeOptions(
  name: string,
  options: Partial<CreateThemeOptions>
): Promise<CreateThemeOptions> {
  while (!name) {
    const { themeName } = await prompts({
      type: 'text',
      name: 'themeName',
      message: 'The name of your theme, for example "Monokai"',
    })
    name = themeName
  }

  let packageName = options.packageName
  if (!packageName) {
    const { themePackageName } = await prompts({
      type: 'text',
      name: 'themePackageName',
      message:
        'The name of the package on npm, for example formkit-theme-monokai.',
      initial: `formkit-theme-${slugify(name)}`,
    })
    packageName = themePackageName as string
  }

  let dirName = options.dir
  let emptyDir = dirName
    ? await isDirEmpty(resolve(process.cwd(), `./${dirName}`))
    : false

  while (!dirName || !emptyDir) {
    const { directory } = await prompts({
      type: 'text',
      name: 'directory',
      message: `The directory to create the theme in.`,
      initial: options.dir || packageName,
    })
    dirName = directory
    emptyDir = await isDirEmpty(resolve(process.cwd(), `./${dirName}`))
  }
  const dir = resolve(process.cwd(), `./${dirName}`)

  return {
    name,
    packageName,
    dir,
  }
}
