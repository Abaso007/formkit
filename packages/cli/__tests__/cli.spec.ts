import { describe, it, expect, vi } from 'vitest'
import { buildTheme } from '../src/buildTheme'
import chalk from 'chalk'
import { readFile } from 'fs/promises'
import { resolve } from 'pathe'

describe('buildTheme', () => {
  it('can build a local theme', () => {
    const consoleMock = vi
      .spyOn(console, 'log')
      .mockImplementation(() => undefined)
    buildTheme('my-theme')
    expect(consoleMock).toHaveBeenCalledWith(
      chalk.greenBright('Locating my-theme...')
    )
    consoleMock.mockRestore()
  })

  it('can generate a local theme', async () => {
    const consoleMock = vi
      .spyOn(console, 'log')
      .mockImplementation(() => undefined)
    await buildTheme('./packages/cli/__tests__/mocks/localTheme', {
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
    })

    expect(consoleMock).toHaveBeenNthCalledWith(
      1,
      chalk.greenBright('Locating ./packages/cli/__tests__/mocks/localTheme...')
    )
    expect(consoleMock).toHaveBeenNthCalledWith(
      2,
      chalk.greenBright('Theme file written to formkit.theme.ts')
    )
    const fileString = await readFile(
      resolve(process.cwd(), 'temp/formkit.theme.ts'),
      'utf-8'
    )
    expect(fileString).toMatchSnapshot()
  })

  it('can override variables in generated theme', async () => {
    await buildTheme('./packages/cli/__tests__/mocks/localTheme', {
      outFile: 'temp/formkit.theme.ts',
      format: 'ts',
      variables: 'border=border-6,spacing=10',
    })
    const fileString = await readFile(
      resolve(process.cwd(), 'temp/formkit.theme.ts'),
      'utf-8'
    )
    expect(fileString).toMatchSnapshot()
  })
})
