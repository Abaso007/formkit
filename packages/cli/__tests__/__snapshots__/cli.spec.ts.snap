// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildTheme > can generate a local theme 1`] = `
"import type { FormKitNode } from '@formkit/core'

/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you’d like to \\"eject\\" from the CLI’s ability to update it.
  *
  * @checksum - a7bfb1d9f983e29f818f08018282b0c2013b3e012133512c2d0559469f51a5a7
  * @variables - undefined
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName: string, node: FormKitNode): Record<string, boolean> {
   const key = \`\${node.props.type}__\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey]
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes: Record<string, Record<string, boolean>> = {
  \\"family:text__outer\\": {
    \\"mb-64\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"text-green-300\\": true
  },
  \\"text__input\\": {
    \\"border-2\\": true,
    \\"rounded-md\\": true,
    \\"p-2\\": true,
    \\"border-slate-50\\": true
  },
  \\"select__inner\\": {
    \\"border-2\\": true,
    \\"rounded-md\\": true,
    \\"p-4\\": true
  },
  \\"select__outer\\": {
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"mb-56\\": true
  }
};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals: Record<string, Record<string, boolean>> = {
  \\"outer\\": {
    \\"mb-64\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true
  }
};
"
`;

exports[`buildTheme > can override variables in generated theme 1`] = `
"import type { FormKitNode } from '@formkit/core'

/**
  * @privateRemarks
  * This file was generated by the FormKit CLI and should not be manually
  * edited unless you’d like to \\"eject\\" from the CLI’s ability to update it.
  *
  * @checksum - 7fe919c13597b626caee472d5179e83221a54e6a8643d8b0073c5bbad94a70e2
  * @variables - border=border-6,spacing=10
  **/

 /**
  * This is the theme function itself, it should be imported and used as the
  * config.rootClasses function. For example:
  *
  * \`\`\`js
  * import { theme } from './formkit.theme'
  * import { defineFormKitConfig } from '@formkit/vue'
  *
  * export default defineFormKitConfig({
  *   config: {
  *     rootClasses: theme
  *   }
  * })
  * \`\`\`
  **/
 export function rootClasses (sectionName: string, node: FormKitNode): Record<string, boolean> {
   const key = \`\${node.props.type}__\${sectionName}\`
   const familyKey = node.props.family ? \`family:\${node.props.family}__\${sectionName}\` : ''
   const memoKey = \`\${key}__\${familyKey}\`
   if (!(memoKey in classes)) {
     const sectionClasses = classes[key] ?? globals[sectionName] ?? {}
     if (familyKey in classes) {
       classes[memoKey] = { ...classes[familyKey],  ...sectionClasses }
     } else {
       classes[memoKey] = sectionClasses
     }
   }
   return classes[memoKey]
 }

/**
 * These classes have already been merged with globals using tailwind-merge
 * and are ready to be used directly in the theme.
 **/
const classes: Record<string, Record<string, boolean>> = {
  \\"family:text__outer\\": {
    \\"mb-10\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"text-green-300\\": true
  },
  \\"text__input\\": {
    \\"border-6\\": true,
    \\"rounded-md\\": true,
    \\"p-2\\": true,
    \\"border-slate-50\\": true
  },
  \\"select__inner\\": {
    \\"border-6\\": true,
    \\"rounded-md\\": true,
    \\"p-4\\": true
  },
  \\"select__outer\\": {
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true,
    \\"mb-8\\": true
  }
};

/**
 * Globals are merged prior to generating this file — these are included for
 * any other non-matching inputs.
 **/
const globals: Record<string, Record<string, boolean>> = {
  \\"outer\\": {
    \\"mb-10\\": true,
    \\"mt-2\\": true,
    \\"ml-80\\": true,
    \\"mr-10\\": true,
    \\"border-green-300\\": true
  }
};
"
`;
